### PR TITLE
FEA Add support for flexiblas

### LINKED ISSUE #154: Consider supporting FlexiBLAS
See https://github.com/scipy/scipy/issues/17362 for details.

### PR DESCRIPTION
**Pull request by jeremiedbb**
currently checking whether it's doable or not

Fixes #154.

### CODE REVIEW

**Comment by ogrisel** [comment]
Thanks for the PR.

Before merging this (assuming we can solver all the technical hurdles), we should wait for the FlexiBLAS maintainers to update their licensing terms as discussed on this scipy thread:

- https://github.com/scipy/scipy/issues/17362#issuecomment-1761202417

File: continuous_integration/posix.yml
Comment (ogrisel) [inline continuous_integration/posix.yml:28]:
```suggestion
        bash continuous_integration/install_with_blis.sh
```

**Comment by jeremiedbb** [comment]
So far, installing flexiblas, linking cython extensions against flexiblas and the flexiblas controller work.
Now I'm trying to build numpy using flexiblas like it's done for BLIS.

**Comment by jeremiedbb** [comment]
```
jeremie@jdb-XPS-13-9360:~/R/O/threadpoolctl$ FLEXIBLAS=BLIS python -m threadpoolctl -i numpy
[
  {
    "user_api": "blas",
    "internal_api": "flexiblas",
    "num_threads": 1,
    "prefix": "libflexiblas",
    "filepath": "/home/jeremie/R/O/flexiblas_install/lib/libflexiblas.so.3.3",
    "version": "3.3.1",
    "available_backends": [
      "NETLIB",
      "MKL",
      "BLIS",
      "OPENBLASPTHREAD",
      "ATLAS"
    ],
    "loaded_backends": [
      "BLIS"
    ],
    "current_backend": "BLIS"
  },
  {
    "user_api": "blas",
    "internal_api": "blis",
    "num_threads": 1,
    "prefix": "libblis",
    "filepath": "/home/jeremie/miniforge/envs/flexiblas_threadpoolctl/lib/libblis.so.4.0.0",
    "version": "0.9.0",
    "threading_layer": "pthreads",
    "architecture": "haswell"
  }
]
```

\o/

File: continuous_integration/install_with_flexiblas.sh
Comment (rgommers) [inline continuous_integration/install_with_flexiblas.sh:56]: You'll want `-Csetup-args=-Dlapack=flexiblas` here as well to link LAPACK. That avoids linking two different libraries for BLAS and LAPACK - from the passing CI log:

```
  Run-time dependency flexiblas found: YES 3.3.1
  Message: BLAS symbol suffix:
  Run-time dependency mkl found: NO (tried pkgconfig and system)
  Run-time dependency accelerate found: NO (tried system)
  Run-time dependency openblas found: YES 0.3.25
```

**Review by ogrisel — COMMENTED** [review]
I gave it a try on my macOS host by building FlexiBLAS and numpy from source by adapting the instruction from the linux CI config of this PR (setting OPENBLASOPENMP as the default BLAS instead of NETLIB) and it seems to work as intended:

```python
>>> import numpy as np
>>> from threadpoolctl import threadpool_info, threadpool_limits
>>> from pprint import pprint
>>> pprint(threadpool_info())
[{'available_backends': ['NETLIB', 'APPLE', 'OPENBLASOPENMP'],
  'current_backend': 'OPENBLASOPENMP',
  'filepath': '/Users/ogrisel/code/flexiblas_install/lib/libflexiblas.3.4.dylib',
  'internal_api': 'flexiblas',
  'loaded_backends': ['OPENBLASOPENMP'],
  'num_threads': 8,
  'prefix': 'libflexiblas',
  'user_api': 'blas',
  'version': '3.4.1'},
 {'architecture': 'VORTEX',
  'filepath': '/Users/ogrisel/miniforge3/envs/flexiblas/lib/libopenblas.0.dylib',
  'internal_api': 'openblas',
  'num_threads': 8,
  'prefix': 'libopenblas',
  'threading_layer': 'openmp',
  'user_api': 'blas',
  'version': '0.3.26'},
 {'filepath': '/Users/ogrisel/miniforge3/envs/flexiblas/lib/libomp.dylib',
  'internal_api': 'openmp',
  'num_threads': 8,
  'prefix': 'libomp',
  'user_api': 'openmp',
  'version': None}]
>>> A = np.random.randn(1024, 1024)
>>> %timeit A @ A.T
12.4 ms ± 1.05 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

and then after setting the `FLEXIBLAS=APPLE` env var:

```python
>>> from threadpoolctl import threadpool_info, threadpool_limits
>>> import numpy as np
>>> from pprint import pprint
>>> pprint(threadpool_info())
[{'available_backends': ['NETLIB', 'APPLE', 'OPENBLASOPENMP'],
  'current_backend': 'APPLE',
  'filepath': '/Users/ogrisel/code/flexiblas_install/lib/libflexiblas.3.4.dylib',
  'internal_api': 'flexiblas',
  'loaded_backends': ['APPLE'],
  'num_threads': 1,
  'prefix': 'libflexiblas',
  'user_api': 'blas',
  'version': '3.4.1'}]
>>> A = np.random.randn(1024, 1024)
>>> %timeit A @ A.T
8.4 ms ± 122 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

`test_flexiblas` fails because it's not possible for me to install blis (at least not via conda-forge on macOS arm64) nor MKL but instead flexiblas can auto-detect the APPLE (Accelerate) backend.

Maybe we should make the `test_flexiblas` test more platform specific and also add CI entries to test flexiblas under Windows and macOS.

Also, it could be nice to have a way to change the flexiblas backend at runtime via threadpoolctl. Even if it's not documented in the FlexiBLAS README, it seems to be possible, for instance:

https://gitlab.mpi-magdeburg.mpg.de/software/flexiblas-release/-/blob/master/examples/demo_api.c?ref_type=heads#L85

File: continuous_integration/install_with_flexiblas.sh
Comment (ogrisel) [inline continuous_integration/install_with_flexiblas.sh:56]: I tried it locally and it seems to work. Let's apply this on the PR to check the updated CI log.

```suggestion
PKG_CONFIG_PATH=$ABS_PATH/numpy/ pip install . -v --no-build-isolation -Csetup-args=-Dblas=flexiblas -Csetup-args=-Dlapack=flexiblas
```

File: threadpoolctl.py
Comment (jeremiedbb) [reply threadpoolctl.py:373]: I don't know when to expect it but I added a comment saying that :smile:

File: tests/test_threadpoolctl.py
Comment (ogrisel) [inline tests/test_threadpoolctl.py:661]: style nitpick:

```suggestion
    if not flexiblas_controller:
```

File: continuous_integration/install_with_flexiblas.sh
Comment (ogrisel) [inline continuous_integration/install_with_flexiblas.sh:31]:
```suggestion
# We intentionally restrict the list of backends to make it easier to
# write platform agnostic tests. In particular, we do not detect OS
# provided backends such as macOS' Apple/Accelerate/vecLib nor plaftorm
# specific BLAS implementations such as MKL that cannot be installed on
# arm64 hardware.
cmake ../ -DCMAKE_INSTALL_PREFIX=$ABS_PATH"/flexiblas_install" \
```

**Review by ogrisel — APPROVED** [review]
Minor comment but otherwise LGTM!

**Pull request merged**
